jetty/src/io/pedestal/http/jetty/websockets.clj

```diff
@@ -42,14 +43,84 @@
   ([on-connect-fn send-buffer-or-n]
    (fn [^Session ws-session]
      (let [send-ch (async/chan send-buffer-or-n)
-           remote ^RemoteEndpoint (.getRemote ws-session)]
+           remote  ^RemoteEndpoint (.getRemote ws-session)]
        ;; Let's process sends...
        (go-loop []
-                (if-let [out-msg (and (.isOpen ws-session)
-                                      (async/<! send-ch))]
-                  (do (ws-send out-msg remote)
-                      (recur))
-                  (.close ws-session)))
+         (if-let [out-msg (and (.isOpen ws-session)
+                               (async/<! send-ch))]
+           (do (try (ws-send out-msg remote)
+                    (catch Exception ex
+                      (log/error :msg "Failed on ws-send"
+                                 :exception ex)))
+               (recur))
+           (.close ws-session)))
+       (on-connect-fn ws-session send-ch)))))
+
+;; Support non-blocking transmission with optional flow control
+(defprotocol WebSocketSendAsync
+  (ws-send-async [msg remote-endpoint]
+    "Sends `msg` to `remote-endpoint`. Returns a
+     promise channel from which the result can be taken."))
+
+(extend-protocol WebSocketSendAsync
+  String
+  (ws-send-async [msg ^RemoteEndpoint remote-endpoint]
+    (let [p-chan (async/promise-chan)]
+      (.sendString remote-endpoint msg (reify WriteCallback
+                                         (writeFailed [_ ex]
+                                           (async/put! p-chan ex))
+                                         (writeSuccess [_]
+                                           (async/put! p-chan :success))))
+      p-chan))
+
+  ByteBuffer
+  (ws-send-async [msg ^RemoteEndpoint remote-endpoint]
+    (let [p-chan (async/promise-chan)]
+      (.sendBytes remote-endpoint msg (reify WriteCallback
+                                        (writeFailed [_ ex]
+                                          (async/put! p-chan ex))
+                                        (writeSuccess [_]
+                                          (async/put! p-chan :success))))
```

Comment:
The reify for  `WriteCallback ` looks 100% the same as the one for String, so I think it could be refactored out (e.g. a deftype?)

Reply:
Yep.